File: 32.최단 경로/1753.최단경로/Main.java
import java.io.*;
import java.util.*;

public class Main {
    static int INF = 20000 * 10;

    public static <T> void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int V = Integer.parseInt(st.nextToken());
        int E = Integer.parseInt(st.nextToken());
        int start = Integer.parseInt(br.readLine());

        List<List<int[]>> graph = new ArrayList<>();
        for (int i = 0; i <= V; i++)
            graph.add(new ArrayList<>());

        for (int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());
            graph.get(u).add(new int[] { v, w });
        }

        int[] distance = new int[V + 1];
        for (int i = 0; i <= V; i++) {
            distance[i] = INF;
        }
        distance[start] = 0;

        Queue<int[]> q = new PriorityQueue<>(new Comparator<int[]>() {
            @Override
            public int compare(int[] o1, int[] o2) {
                return o1[1] - o2[1];
            }
        });

        q.offer(new int[] { start, 0 });

        while (q.isEmpty() == false) {
            int[] cur = q.poll();
            int curNode = cur[0];
            int curDist = cur[1];

            if (curDist > distance[curNode])
                continue;

            for (int[] next : graph.get(curNode)) {
                int nextNode = next[0];
                int needDist = next[1];
                int nextDist = curDist + needDist;

                if (nextDist < distance[nextNode]) {
                    distance[nextNode] = nextDist;
                    q.offer(new int[] { nextNode, nextDist });
                }
            }
        }

        for (int i = 1; i <= V; i++) {
            if (distance[i] == INF) {
                bw.write("INF\n");
                continue;
            }
            bw.write(String.valueOf(distance[i]) + "\n");
        }

        br.close();
        bw.close();
    }

}

File: 32.최단 경로/1753.최단경로/main.py
import sys
import heapq

input = sys.stdin.readline
INF = 20000 * 10

if __name__ == "__main__":
    # 정점 개수, 간선 개수 입력 받기
    V, E = map(int, input().split())
    # 시작점 입력 받기
    start = int(input())
    # 각 노드에 연결된 노드와 경로 입력받기
    graph = [[] for _ in range(V + 1)]  # 1 ~ V
    for _ in range(E):
        u, v, w = map(int, input().split())
        graph[u].append((v, w))

    # 시작점에서 준비
    heap = []
    distance = [INF] * (V + 1)
    distance[start] = 0
    heapq.heappush(heap, (0, start))

    # 연산 시작
    while heap:
        # 현재 노드까지의 거리, 현재 노드
        cur_dist, cur_node = heapq.heappop(heap)
        # 현재 노드가 이미 처리된 적이 있을 때 continue
        if cur_dist > distance[cur_node]:
            continue

        # 현재 노드와 연결된 다른 인접 노드들을 확인
        for next_node, need_dist in graph[cur_node]:
            # 다음 노드까지의 거리
            cost = cur_dist + need_dist
            if cost < distance[next_node]:
                distance[next_node] = cost
                heapq.heappush(heap, (cost, next_node))

    for i in range(1, V + 1):
        print("INF" if distance[i] == INF else distance[i])
